dkg/src/epoch_manager.rs

```diff
@@ -0,0 +1,207 @@
+        }
+    }
+
+    async fn await_reconfig_notification(&mut self) {
+        let reconfig_notification = self
+            .reconfig_events
+            .next()
+            .await
+            .expect("Reconfig sender dropped, unable to start new epoch");
+        self.start_new_epoch(reconfig_notification.on_chain_configs)
+            .await;
+    }
+
+    async fn start_new_epoch(&mut self, payload: OnChainConfigPayload<P>) {
+        let validator_set: ValidatorSet = payload
+            .get()
+            .expect("failed to get ValidatorSet from payload");
+
+        let epoch_state = EpochState {
+            epoch: payload.epoch(),
+            verifier: (&validator_set).into(),
+        };
+        self.epoch_state = Some(Arc::new(epoch_state.clone()));
+
+        let features = payload.get::<Features>().unwrap_or_default();
+
+        if features.is_enabled(FeatureFlag::RECONFIGURE_WITH_DKG) {
+            let DKGState {
+                in_progress: in_progress_session,
+                ..
+            } = payload.get::<DKGState>().unwrap_or_default();
+
+            let agg_node_producer = DummyAggNodeProducer {}; //TODO: replace with real
+
+            let (start_dkg_event_tx, start_dkg_event_rx) =
+                aptos_channel::new(QueueStyle::KLAST, 1, None);
+            self.start_dkg_event_tx = Some(start_dkg_event_tx);
+
+            let (dkg_rpc_msg_tx, dkg_rpc_msg_rx) = aptos_channel::new::<
+                (),
+                (AccountAddress, IncomingRpcRequest),
+            >(QueueStyle::FIFO, 100, None);
+            self.dkg_rpc_msg_tx = Some(dkg_rpc_msg_tx);
+            let (dkg_manager_close_tx, dkg_manager_close_rx) = oneshot::channel();
+            self.dkg_manager_close_tx = Some(dkg_manager_close_tx);
+
+            let dkg_manager = DKGManager::new(
+                self.my_addr,
+                self.epoch_state().clone(),
+                Arc::new(agg_node_producer),
+                self.vtxn_pool_write_cli.clone(),
+            );
+            let (vtxn_pull_notification_tx, vtxn_pull_notification_rx) =
+                aptos_channel::new(QueueStyle::KLAST, 1, None);
+            self.vtxn_pull_notification_tx_to_dkgmgr = Some(vtxn_pull_notification_tx);
+            tokio::spawn(dkg_manager.run(
+                in_progress_session,
+                start_dkg_event_rx,
+                dkg_rpc_msg_rx,
+                vtxn_pull_notification_rx,
+                dkg_manager_close_rx,
+            ));
+        }
+    }
+
+    async fn on_new_epoch(&mut self, reconfig_notification: ReconfigNotification<P>) -> Result<()> {
+        self.shutdown_current_processor().await;
+        self.start_new_epoch(reconfig_notification.on_chain_configs)
+            .await;
+        Ok(())
+    }
+
+    async fn shutdown_current_processor(&mut self) {
+        if let Some(tx) = self.dkg_manager_close_tx.take() {
+            let (ack_tx, ack_rx) = oneshot::channel();
+            tx.send(ack_tx).unwrap();
```

Comment:
unwrap is safe, right? This is the only way to shutdown dkg manager?

Reply:
i think it's safe. And yes that's the only way today.